aptos-move/framework/aptos-framework/sources/scheduled_transactions/scheduled_txns.move

```diff
@@ -0,0 +1,1171 @@
+    public fun insert(
+        sender: &signer, txn: ScheduledTransaction
+    ): ScheduleMapKey acquires ScheduleQueue, AuxiliaryData {
+        // If scheduling is shutdown, we cannot schedule any more transactions
+        let aux_data = borrow_global<AuxiliaryData>(@aptos_framework);
+        assert!(
+            (aux_data.module_status == ScheduledTxnsModuleStatus::Active),
+            error::unavailable(EUNAVAILABLE)
+        );
+
+        // we expect the sender to be a permissioned signer
+        assert!(
+            signer::address_of(sender) == txn.sender_addr,
+            error::permission_denied(EINVALID_SIGNER)
+        );
+
+        // Only schedule txns in the future
+        let txn_time = txn.scheduled_time_ms;
+        let block_time = timestamp::now_microseconds() / 1000;
+        assert!(txn_time > block_time, error::invalid_argument(EINVALID_TIME));
+
+        assert!(
+            txn.max_gas_unit_price >= 100,
+            error::invalid_argument(ELOW_GAS_UNIT_PRICE)
+        );
+
+        assert!(
+            bcs::serialized_size(&txn) < MAX_SCHED_TXN_SIZE,
+            error::invalid_argument(ETXN_TOO_LARGE)
+        );
+
+        // Generate unique transaction ID
+        let hash = sha3_256(bcs::to_bytes(&txn));
+        let txn_id = hash_to_u256(hash);
+
+        // Insert the transaction into the schedule_map
+        // Create schedule map key
+        let key = ScheduleMapKey {
+            time: txn_time,
+            gas_priority: U64_MAX - txn.max_gas_unit_price,
+            txn_id
+        };
+
+        let queue = borrow_global_mut<ScheduleQueue>(@aptos_framework);
+        queue.schedule_map.add(key, Empty {});
+        queue.txn_tbl.add(key, txn);
+
+        // Collect deposit
+        // Get owner signer from capability
+        let gas_deposit_store_cap = borrow_global<AuxiliaryData>(@aptos_framework);
+        let gas_deposit_store_addr =
+            account::get_signer_capability_address(
+                &gas_deposit_store_cap.gas_fee_deposit_store_signer_cap
+            );
+
+        coin::transfer<AptosCoin>(
+            sender,
+            gas_deposit_store_addr,
+            txn.max_gas_amount * txn.max_gas_unit_price
+        );
+
+        // Emit event that txn has been scheduled; for now indexer wants to consume this
+        event::emit(
+            TransactionScheduledEvent {
+                scheduled_txn_time: txn_time,
+                scheduled_txn_hash: txn_id,
+                sender_addr: txn.sender_addr,
+                scheduled_time_ms: txn.scheduled_time_ms,
+                max_gas_amount: txn.max_gas_amount,
+                max_gas_unit_price: txn.max_gas_unit_price
+            }
+        );
+
+        key
+    }
+
+    /// Cancel a scheduled transaction, must be called by the signer who originally scheduled the transaction.
+    public fun cancel(sender: &signer, key: ScheduleMapKey) acquires ScheduleQueue, AuxiliaryData {
```

Comment:
do we want to put some constrain on the cancel (e.g. you cannot cannel a txn that is going to be run shortly)